### PR TITLE
Replay slots from blocktree in new forks

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

Replay stage is not tracking forks correctly as they are delivered to Blocktree.

#### Summary of Changes

Replay stage main loop:
1. for all frozen forks, ask blocktree for any new slots chaining to those forks, and create new banks
2. for all live forks, ask blocktree for any new blobs for those forks
3. if any live forks process to a max tick for the slot, vote
4. goto 1

Fixes #

### CODE REVIEW

File: src/bank_forks.rs [hunk 1]
```diff
@@ -18,11 +18,28 @@ impl BankForks {
...
     }
-
+    pub fn frozen_banks(&self) -> Vec<u64> {
+       self.banks
+            .iter()
+            .filter(|(k, v)| v.is_frozen())
+            .map(|x| *x.0)
+            .collect()
+    }
+    pub fn live_banks(&self) -> Vec<u64> {
```
Comment (rob-solana) [inline src/bank_forks.rs]: the banks in BankForks should all be live...  blocktree_processor populates BankForks with children, and insert() removes parents...

File: src/tvu.rs [hunk 2]
```diff
@@ -34,7 +33,7 @@ use std::sync::{Arc, RwLock};
 use std::thread;
 
 pub struct TvuRotationInfo {
-    pub bank: Arc<Bank>,     // Bank to use
```
Comment (rob-solana) [inline src/tvu.rs]: you wouldn't rather just take a bank and a slot?

File: src/replay_stage.rs [hunk 3]
```diff
@@ -270,146 +90,71 @@ impl ReplayStage {
...
-                            cluster_info.write().unwrap().set_leader(leader_id);
+                        if next_leader == my_id {
+                            let tpu_bank = Bank::new_from_parent_and_id(bank, my_id, next_slot, true);
+                            bank_forks.write().unwrap().insert(next_slot, tpu_bank);
+
+                            // Always send rotation signal so that other services like
+                            // RPC can be made aware of last slot's bank
+                            to_leader_sender
+                                .send(TvuRotationInfo {
+                                    tick_height: bank.tick_height(),
```
Comment (rob-solana) [inline src/replay_stage.rs]: how come we don't just send the new Bank?

File: runtime/src/bank.rs [hunk 4]
```diff
@@ -111,6 +111,9 @@ pub struct Bank {
 
     /// The pubkey to send transactions fees to.
     collector_id: Pubkey,
+
+    /// if this is a tpu bank
+    pub tpu: bool,
```
Comment (garious) [inline runtime/src/bank.rs]: This seems way too context aware. Bank can be used by anything.  Take, for example, the field just above it. It's called a "collector", not a "leader". If we must have this boolean, can it instead be in the form `is_<capability>`?

File: src/bank_forks.rs (hunk 1 above)
Comment (garious) [reply src/bank_forks.rs]: This is where we were using the term `active` in the proposal, not `live`.

File: src/bank_forks.rs [hunk 5]
```diff
@@ -18,11 +18,28 @@ impl BankForks {
...
+            .iter()
+            .filter(|(k, v)| !v.is_frozen())
+            .map(|x| *x.0)
+            .collect()
+    }
     pub fn working_bank(&self) -> Arc<Bank> {
         self.banks[&self.working_bank_id].clone()
     }
 
+    pub fn lookup(&self, bank_id: u64) -> Option<&Arc<Bank>> {
```
Comment (garious) [inline src/bank_forks.rs]: `get` is a fine name. "when in Rome..."

File: src/replay_stage.rs (hunk 3 above)
Comment (aeyakovenko) [reply src/replay_stage.rs]: @rob-solana because the notification will move into replay stage.

File: runtime/src/bank.rs (hunk 4 above)
Comment (aeyakovenko) [reply runtime/src/bank.rs]: @garious what about sticking into forks?

Comment (garious) [reply runtime/src/bank.rs]: Sure, but that's the working bank, which I thought we were working to purge.

Comment (aeyakovenko) [reply runtime/src/bank.rs]: @garious, I was thinking as metadata for the banks.

Comment (garious) [reply runtime/src/bank.rs]: Sounds like you need a `BankEntry` struct in `BankForks` that wraps `Bank` and adds metadata.

File: src/replay_stage.rs [hunk 6]
```diff
@@ -270,146 +116,101 @@ impl ReplayStage {
...
+                        if next_leader == my_id {
+                            let tpu_bank =
+                                Bank::new_from_parent_and_id(&bank, my_id, next_slot);
+                            bank_forks.write().unwrap().insert(next_slot, tpu_bank);
                         }
-
                         // Always send rotation signal so that other services like
                         // RPC can be made aware of last slot's bank
                         to_leader_sender
                             .send(TvuRotationInfo {
```
Comment (carllin) [inline src/replay_stage.rs]: We could potentially send multiple instances of these signals while an older TPU is still running right?

Comment (aeyakovenko) [reply src/replay_stage.rs]: Yea. I don’t think it matters.

Comment (carllin) [reply src/replay_stage.rs]: Yeah I think worst case that older fork just stops making progress, and then we can purge it later

**Comment by rob-solana** [comment]
looks tight

File: src/tvu.rs (hunk 2 above)
Comment (aeyakovenko) [reply src/tvu.rs]: @rob-solana block tree doesn't have any data for the next bank yet.

File: src/replay_stage.rs (hunk 3 above)
Comment (aeyakovenko) [reply src/replay_stage.rs]: @rob-solana where is the new bank?  i just have a parent in that context.

File: src/replay_stage.rs [hunk 7]
```diff
@@ -180,222 +71,169 @@ impl ReplayStage {
...
-                            if let Ok(entries) = blocktree.get_slot_entries(
-                                slot,
-                                current_blob_index,
-                                Some(MAX_ENTRY_RECV_PER_ITER as u64),
-                            ) {
-                                entries
-                            } else {
-                                vec![]
+                        let max_tick_height = (bank_id + 1) * bank.ticks_per_slot() - 1;
+                        if bank.tick_height() == max_tick_height {
```
Comment (rob-solana) [inline src/replay_stage.rs]: if it's > max_tick_height, we do what?

File: src/replay_stage.rs [hunk 8]
```diff
@@ -180,222 +71,169 @@ impl ReplayStage {
...
+                            .read()
+                            .unwrap()
+                            .get(*latest_slot_vote)
+                            .unwrap()
+                            .clone();
+                        let next_slot = *latest_slot_vote + 1;
+                        let next_leader = leader_schedule_utils::slot_leader_at(next_slot, &parent);
+                        cluster_info.write().unwrap().set_leader(next_leader);
+
+                        subscriptions.notify_subscribers(&parent);
```
Comment (rob-solana) [inline src/replay_stage.rs]: awesome

File: src/replay_stage.rs [hunk 9]
```diff
@@ -180,222 +71,169 @@ impl ReplayStage {
...
-                        } else {
-                            current_slot = None;
+                            cluster_info.write().unwrap().push_vote(vote);
                         }
-
-                        if leader_id != current_leader_id {
-                            // TODO: Remove this soon once we boot the leader from ClusterInfo
-                            cluster_info.write().unwrap().set_leader(leader_id);
+                        if next_leader == my_id {
+                            let mut wforks = bank_forks.write().unwrap();
```
Comment (rob-solana) [inline src/replay_stage.rs]: can you make this a one-liner?  these wXXX or rXXX bug my eyes...

`bank_forks.write().unwrap().insert(tpu_bank);`

File: src/replay_stage.rs (hunk 7 above)
Comment (carllin) [reply src/replay_stage.rs]: that means its a bad fork and will eventually get timed out/pruned?

File: src/replay_stage.rs [hunk 10]
```diff
@@ -180,222 +71,169 @@ impl ReplayStage {
...
 
+    pub fn replay_blocktree_into_bank(
+        bank: &Bank,
+        blocktree: &Blocktree,
+        progress: &mut HashMap<u64, (Hash, usize)>,
+        forward_entry_sender: &EntrySender,
+    ) -> result::Result<()> {
+        let (entries, num) = Self::load_blocktree_entries(bank, blocktree, progress)?;
+        let len = entries.len();
+        let result =
```
Comment (rob-solana) [inline src/replay_stage.rs]: separation of blob count and entry count, yay!

File: src/replay_stage.rs [hunk 11]
```diff
@@ -180,222 +71,169 @@ impl ReplayStage {
...
         )
     }
 
+    pub fn replay_blocktree_into_bank(
+        bank: &Bank,
+        blocktree: &Blocktree,
+        progress: &mut HashMap<u64, (Hash, usize)>,
+        forward_entry_sender: &EntrySender,
+    ) -> result::Result<()> {
+        let (entries, num) = Self::load_blocktree_entries(bank, blocktree, progress)?;
```
Comment (rob-solana) [inline src/replay_stage.rs]: separation of blob count and entry count, yay!

**Comment by garious** [comment]
Can you rebase?

**Pull request merged**